retrieval/discovery/gce.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"google.golang.org/api/compute/v1"
+
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	gceLabel             = model.MetaLabelPrefix + "gce_"
+	gceLabelProject      = gceLabel + "project"
+	gceLabelZone         = gceLabel + "zone"
+	gceLabelNetwork      = gceLabel + "network"
+	gceLabelSubnetwork   = gceLabel + "subnetwork"
+	gceLabelPublicIP     = gceLabel + "public_ip"
+	gceLabelPrivateIP    = gceLabel + "private_ip"
+	gceLabelInstanceName = gceLabel + "instance_name"
+	gceLabelTags         = gceLabel + "tags"
+)
+
+// GCEDiscovery periodically performs GCE-SD requests. It implements
+// the TargetProvider interface.
+type GCEDiscovery struct {
+	project      string
+	zone         string
+	filter       string
+	client       *http.Client
+	svc          *compute.Service
+	isvc         *compute.InstancesService
+	interval     time.Duration
+	port         int
+	tagSeparator string
+}
+
+// NewGCEDiscovery returns a new GCEDiscovery which periodically refreshes its targets.
+func NewGCEDiscovery(conf *config.GCESDConfig) *GCEDiscovery {
+	gd := &GCEDiscovery{
+		project:      conf.Project,
+		zone:         conf.Zone,
+		filter:       conf.Filter,
+		interval:     time.Duration(conf.RefreshInterval),
+		port:         conf.Port,
+		tagSeparator: conf.TagSeparator,
+	}
+	var err error
+	gd.client, err = google.DefaultClient(oauth2.NoContext, compute.ComputeReadonlyScope)
+	if err != nil {
+		log.Errorf("error setting up communication with GCE service: %s", err)
+		return gd
+	}
+	gd.svc, err = compute.New(gd.client)
+	if err != nil {
+		log.Errorf("error setting up communication with GCE service: %s", err)
+		return gd
+	}
+	gd.isvc = compute.NewInstancesService(gd.svc)
+	return gd
+}
+
+// Run implements the TargetProvider interface.
+func (gd *GCEDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(gd.interval)
```

Comment:
Any reason to not move the ticker initialization down above the for loop? No need to start one if we error out early.